src/pyupgrade/rules/format_literals.rs

```diff
@@ -0,0 +1,118 @@
+    let mut new_args: Vec<Arg> = Vec::with_capacity(old_args.len());
+    for (idx, given) in correct_order.iter().enumerate() {
+        // We need to keep the formatting in the same order but move the values.
+        let values = old_args
+            .get(*given)
+            .ok_or_else(|| anyhow!("Failed to extract argument at: {given}"))?;
+        let formatting = old_args
+            .get(idx)
+            .ok_or_else(|| anyhow!("Failed to extract argument at: {idx}"))?;
+        let new_arg = Arg {
+            value: values.value.clone(),
+            comma: formatting.comma.clone(),
+            equal: None,
+            keyword: None,
+            star: values.star,
+            whitespace_after_star: formatting.whitespace_after_star.clone(),
+            whitespace_after_arg: formatting.whitespace_after_arg.clone(),
+        };
+        new_args.push(new_arg);
+    }
+    Ok(new_args)
+}
+
+/// Returns the corrected function call.
+fn generate_call(module_text: &str, correct_order: &[usize]) -> Result<String> {
+    let mut expression = match_expression(module_text)?;
+    let mut call = match_call(&mut expression)?;
+
+    // Fix the call arguments.
+    call.args = generate_arguments(&call.args, correct_order)?;
+
+    // Fix the string itself.
+    let Expression::Attribute(item) = &*call.func else {
+        panic!("Expected: Expression::Attribute")
+    };
+
+    let mut state = CodegenState::default();
+    item.codegen(&mut state);
+    let cleaned = remove_specifiers(&state.to_string());
+
+    call.func = Box::new(match_expression(&cleaned)?);
+
+    let mut state = CodegenState::default();
+    expression.codegen(&mut state);
+    if module_text == state.to_string() {
+        // Ex) `'{' '0}'.format(1)`
+        bail!("Failed to generate call expression for: {module_text}")
+    }
+    Ok(state.to_string())
+}
+
+/// UP030
+pub(crate) fn format_literals(checker: &mut Checker, summary: &FormatSummary, expr: &Expr) {
+    // The format we expect is, e.g.: `"{0} {1}".format(...)`
+    if summary.has_nested_parts {
+        return;
+    }
+    if !summary.keywords.is_empty() {
+        return;
+    }
+    if !summary.autos.is_empty() {
+        return;
+    }
+    if !(0..summary.indexes.len()).all(|index| summary.indexes.contains(&index)) {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(violations::FormatLiterals, Range::from_located(expr));
+    if checker.patch(diagnostic.kind.code()) {
+        // Currently, the only issue we know of is in LibCST:
+        // https://github.com/Instagram/LibCST/issues/846
+        if let Ok(contents) = generate_call(
+            &checker
+                .locator
+                .slice_source_code_range(&Range::from_located(expr)),
+            &summary.indexes,
+        ) {
+            diagnostic.amend(Fix::replacement(
+                contents,
+                expr.location,
+                expr.end_location.unwrap(),
+            ));
```

Comment:
I feel like one strategy that could work here would be...

- Take the entire text.
- Replace the string part via the regex above.
- Replace the arguments via LibCST.

Kinda tough, hacky, etc... but would solve the missing cases.

Reply:
Actually, I guess that wouldn't solve the `'{' '0}'.format(1)` case.